File: Programmers/Level 2/Level2_시소짝꿍.py
weights = [100,180,360,100,270]

result = 0
weights.sort()
for i in range(len(weights)):
    for j in range(i + 1, len(weights)):
        if weights[i] == weights[j]:
            result += 1
        if weights[j] == weights[i] * 2:
            result += 1
        if weights[j] == weights[i] * 3 / 2:
            result += 1
        if weights[j] == weights[i] * 4 / 3:
            result += 1
print(result)
'''
w[i] * (2, 3, 4) == w[j] * (2, 3, 4)
            a                   b
=> w[j] = w[i] * a / b
'''

# list의 길이가 10만이기에 2중 포문 시 시간 초과 발생했음.

# 다른 사람 풀이
# 단일 반복문으로 결과를 도출할 수 있도록 했음
def solution(weights):
    answer = 0    
    dic = {}    

    # 동일한 몸무게의 숫자 확인
    for i in weights:
        if i in dic:
            dic[i] += 1
        else:
            dic[i] = 1

    for i in dic:        
        # 동일한 몸무게의 가능한 조합의 수
        if dic[i] > 1:
            answer += (dic[i] * (dic[i]-1)) / 2
        
        # 각 시소의 거리에 대하여 dic에 존재하는지를 확인하여 쌍 확인
        if i*2 in dic:
            answer += dic[i] * dic[2*i]
        if i*2/3 in dic:
            answer += dic[i] * dic[i*2/3]
        if i*3/4 in dic:
            answer += dic[i] * dic[i*3/4]    
    
    return answer

print(solution(weights))

File: Programmers/Level 2/Level2_시소짝꿍.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <queue>
#include <stack>
#include <map>
using namespace std;

long long solution(vector<int> weights) {
    long long answer = 0;

    // map을 사용하여 중복 제거 및 카운팅
    map<long long, long long> weightMap;
    for (int i = 0; i < weights.size(); i++)
        weightMap[weights[i]] += 1;

    // a == a, 2a == 3b  2a == 4b  3a == 4b
    for (auto iter = weightMap.begin(); iter != weightMap.end(); iter++)
    {
        // 같은 n개중 2개 선택 조합
        if (iter->second > 1)
            answer += (iter->second * (iter->second - 1)) / 2;

        for (auto iter2 = next(iter, 1); iter2 != weightMap.end(); iter2++)
        {
            if (2 * iter->first == 3 * iter2->first || 3 * iter->first == 2 * iter2->first)
                answer += iter->second * iter2->second;
            if (2 * iter->first == 4 * iter2->first || 4 * iter->first == 2 * iter2->first)
                answer += iter->second * iter2->second;
            if (3 * iter->first == 4 * iter2->first || 4 * iter->first == 3 * iter2->first)
                answer += iter->second * iter2->second;
        }
    }

    return answer;
}

// 1차 시도 : 실패, 테스트케이스 12~15

// ㅈㅈ.. 아무리 생각해도 모르겠어서 질문목록 중 나와 동일하게 실패한 사람을 확인해본 결과
// int의 범위값을 벗어나는 문제였다. map<int, int>에서 map<long long, long long> 변경 후 성공
